### PR TITLE
merge common.diag and dense.diag

### PR DESCRIPTION
**Pull request by smichr**
#### Brief description of what is fixed or changed
- Matrix([1, [2]]) now raises an error
- diag shape defaults to square if only rows or cols is given
- keywords allow control of Matrix.diag since dense.diag behaved different than common.diag and the two are now merged.

The publicly facing `diag` and private `Matrix.diag` now have controls:

1. a single list can be unpacked instead of being converted to a matrix with keyword `unpack`
```python
>>> diag([1, 2, 3])
Matrix([
[1],
[2],
[3]])
>>> diag([1, 2, 3], unpack=True) # = diag(*[1, 2, 3]) = diag(1, 2, 3) = Matrix.diag([1, 2, 3])
Matrix([
[1, 0, 0],
[0, 2, 0],
[0, 0, 3]]
```

2. The conversion of lists to matrices can be relaxed so list of ragged rows can be accepted:
```python
>>> diag([[1, 2, 3], 4], strict=False)
Matrix([
[1, 2, 3],
[4, 0, 0]])
```
This is the default behavior of Matrix.diag (but unpack must be False for such a single list or an additional level of brackets must be added)
```python
>>> Matrix.diag([[1, 2, 3], 4], unpack=False)
Matrix([
[1, 2, 3],
[4, 0, 0]])
>>> Matrix.diag([  [[1, 2, 3], 4]  ])
Matrix([
[1, 2, 3],
[4, 0, 0]])
```

(Such input, formerly, put list-literals into the resulting matrix, making it non-functional.)

#### Other comments

During previous cleanup of the matrix files structure, two versions
of `diag` were left behind. The one that is public facing via
`from sympy import *` is the one in `dense.py`. It differs from
that in `common.py` a.k.a `Matrix.diag` by converting an input list to a matrix
rather than unpacking it. Hence, it is strict about matrix structure.

The `Matrix.diag` was not strict and actually could produce non-functional
matrices. An aspect of non-strictness is retained while doing the right thing
more often; an option to strictly convert lists to matrices is controlled with
the keyword `strict`. It is True for the public `diag` and False for the private `Matrix.diag`.

#### Release Notes

- matrices
    - matrix creation detects singleton lists in matrices so Matrix([1, [2]]) now raises an error
    - diag matrix shape defaults to square if only rows or cols is given
    - diag unpacking of lists (conversion of single lists to matrices) can be controlled with keyword `unpack`; dense.diag defaults to True while Matrix.diag (common.diag) defaults to False; `Matrix.diag(*[1, 2, 3]) == Matrix.diag([1, 2, 3], unpack=True)`)
    - diag accepts variable-length rows if keyword `strict` is set to False

### CODE REVIEW

**Comment by sylee957** [comment]
The merging would look good.

But I have some questions about this implementation.



From first and second examples, I find that this change is backward incompatible.

And from the third one, is it valid?

**Comment by smichr** [comment]
I pushed a commit that retains the unpacking difference between the two. I think it would be good to deprecate this difference, however.

I'm not sure what to do about the "fill diagonal" vs "fill from last diagonal". Like I said, I think this is an oversight and an undesirable characteristic that would only affect filling with non-square elements.

**Comment by smichr** [comment]
@asmeurer , could you give a thumbs up or down with making the following change without deprecation? I think the previous behavior does not make sense -- is more a bug than a feature -- for a method named "diag" which should be filling items in on the diagonal:
```python
>>> diag([1, 2, 3], [4, 5], 6)
Matrix([
[1, 0, 0],
[2, 4, 0],
[3, 5, 6]])

FORMERLY

Matrix([
[1, 0, 0],
[2, 0, 0],
[3, 0, 0],
[0, 4, 0],
[0, 5, 0],
[0, 0, 6]])
```

**Comment by smichr** [comment]
I've got to think about how to combine these better.

File: sympy/matrices/dense.py [hunk 1]
```diff
@@ -1103,103 +1103,42 @@ def eye(*args, **kwargs):
...
-    values = (normalize(m) for m in values)
-
-    return Matrix.diag(*values, **kwargs)
+    if 'unpack' not in kwargs:
+        kwargs['unpack'] = True
+        rv = Matrix.diag(*values, **kwargs)
+        kwargs.pop('unpack')
+        return rv
+    else:
+        return Matrix.diag(*values, **kwargs)
```
[the remark this answers is not in the document]
Comment (smichr) [reply sympy/matrices/dense.py]: Oh, that's the way to add an additional kword. Thanks!

**Comment by asmeurer** [comment]
Is there a motivation for the breaking change? I'm not really sure which behavior would be more useful.  

The equivalent I found for NumPy is `scipy.linalg.block_diag`, which works like `sympy.diag` does in master https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.linalg.block_diag.html. 

Regarding the history of this, there was originally `block_diag`, which only supported square matrices. Then `diag` was added in 25212c818dfaaef8db4eb43bb951dd4b91afa62c, which supported any shape (the behavior in master). Eventually `block_diag` was deprecated and removed in favor of `diag`.

**Comment by asmeurer** [comment]
The current behavior of diag is how I would read a block matrix with non-square blocks on the diagonal. Like https://en.wikipedia.org/wiki/Block_matrix#Block_diagonal_matrices if the A1, A2, ... were not square. 

I think we should keep the current `diag` behavior. I think it is useful (at least since it is how scipy's block_diag works), and it has been there for a while. We can add the other behavior under a flag, although I'm still unclear when exactly you would use it.

**Comment by smichr** [comment]
Very helpful history, @asmeurer. Thanks. I missed the deprecation of block_diag. It makes sense now to keep the old behavior. A more general `banded` function in progress will be able to create matrices which are described by the elements on the diagonals.

I also see how the notion of block would make the fill-from-last- diagonal make sense...if you don't then you aren't describing blocks.

File: sympy/matrices/dense.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jksuom) [reply sympy/matrices/dense.py]: That will not work as long as Python 2.7 is supported.

[the remark this answers is not in the document]
Comment (asmeurer) [reply sympy/matrices/dense.py]: I think it works. The problem in 2.7 is doing that in the parameter list of a function definition.

**Comment by smichr** [comment]
Commit history will be cleaned up and I will make any additional changes that are requested. @sylee957 , are you able to give this a look again? There is now no backwards incompatible changes.

**Comment by smichr** [comment]
- [ ] note that diag of non-square matrices is just the "direct sum" referenced [here](https://en.wikipedia.org/wiki/Block_matrix)

**Pull request merged**